runtime/drivers/duckdb/duckdb.go

```diff
@@ -206,18 +234,78 @@ func (c *connection) acquireOLAPConn(ctx context.Context, priority int) (*sqlx.C
 	}
 
 	// Get new conn
-	conn, err = c.db.Connx(ctx)
+	conn, releaseConn, err := c.acquireConn(ctx)
 	if err != nil {
 		c.olapSem.Release()
 		return nil, nil, err
 	}
 
 	// Build release func
 	release := func() error {
-		err := conn.Close()
+		err := releaseConn()
 		c.olapSem.Release()
 		return err
 	}
 
 	return conn, release, nil
 }
+
+// acquireConn returns a DuckDB connection. It should only be used internally in acquireMetaConn and acquireOLAPConn.
+// acquireConn implements the connection tracking and DB reopening logic described in the struct definition for connection.
+func (c *connection) acquireConn(ctx context.Context) (*sqlx.Conn, func() error, error) {
+	c.dbCond.L.Lock()
+	for {
+		if c.dbErr != nil {
+			c.dbCond.L.Unlock()
+			return nil, nil, c.dbErr
+		}
+		if !c.dbReopen {
+			break
+		}
+		c.dbCond.Wait()
+	}
+
+	c.dbConnCount++
+	c.dbCond.L.Unlock()
+
+	conn, err := c.db.Connx(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	release := func() error {
+		err := conn.Close()
+		c.dbCond.L.Lock()
+		c.dbConnCount--
+		if c.dbConnCount == 0 && c.dbReopen {
+			c.dbReopen = false
+			err = c.reopenDB()
+			if err == nil {
+				c.logger.Info("reopened DuckDB successfully")
+			} else {
+				c.logger.Error("reopen of DuckDB failed - the handle is now permanently locked", zap.Error(err))
+			}
+			c.dbErr = err
+			c.dbCond.Broadcast()
+		}
+		c.dbCond.L.Unlock()
+		return err
+	}
+
+	return conn, release, nil
+}
+
+// checkErr marks the DB for reopening if the error is an internal DuckDB error.
+// In all other cases, it just proxies the err.
+// It should be wrapped around errors returned from DuckDB queries. **It must be called while still holding an acquired DuckDB connection.**
+func (c *connection) checkErr(err error) error {
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "INTERNAL Error:") || strings.HasPrefix(err.Error(), "FATAL Error") {
+			c.dbCond.L.Lock()
+			c.dbReopen = true
+			c.logger.Error("encountered internal DuckDB error - scheduling reopen of DuckDB", zap.Error(err))
+			c.dbCond.L.Unlock()
```

Comment:
Use `defer unlock` construct unless there is some specific reason to use this.

Reply:
Done